src/sink.rs

```diff
@@ -0,0 +1,847 @@
+                    }
+                    Error(err) => {
+                        // drop the publish stream's actual element, as if the publish had failed
+                        std::mem::drop(self.messages.next());
+
+                        // replace it with this error
+                        Poll::Ready(Some(Err(err)))
+                    }
+                }
+            }
+        }
+
+        #[derive(Debug)]
+        struct ControlledPublisher {
+            command: Rc<RefCell<PublishCommand>>,
+            mock: MockPublisher,
+        }
+
+        impl ControlledPublisher {
+            fn new() -> (Self, Rc<RefCell<PublishCommand>>) {
+                let command = Rc::new(RefCell::new(PublishCommand::Pending));
+                (
+                    ControlledPublisher {
+                        command: Rc::clone(&command),
+                        mock: MockPublisher::new(),
+                    },
+                    command,
+                )
+            }
+        }
+
+        impl Publisher for ControlledPublisher {
+            type MessageId = ();
+            type MessageError = TestError;
+            type PublishStream = ControlledPublishStream;
+
+            fn publish<'a, I>(&self, topic: Topic, messages: I) -> Self::PublishStream
+            where
+                I: Iterator<Item = &'a ValidatedMessage> + ExactSizeIterator + DoubleEndedIterator,
+            {
+                ControlledPublishStream {
+                    command: Rc::clone(&self.command),
+                    mock: self.mock.clone(),
+                    messages: messages
+                        .cloned()
+                        .map(|msg| (topic, msg))
+                        .collect::<Vec<_>>()
+                        .into_iter(),
+                }
+            }
+        }
+
+        /// The publisher should check that a batch size of zero doesn't crash and instead uses
+        /// something like a max(1, size)
+        #[test]
+        fn batch_size_zero() {
+            let batch_size = 0;
+            let publisher = MockPublisher::new();
+            let sink = publisher_sink(publisher, batch_size);
+            pin_mut!(sink);
+            let mut cx = Context::from_waker(futures_util::task::noop_waker_ref());
+
+            assert_eq!(Poll::Ready(Ok(())), sink.as_mut().poll_ready(&mut cx));
+            assert_eq!(
+                Ok(()),
+                sink.as_mut()
+                    .start_send(("test_topic", test_validated_message("foo")))
+            );
+        }
+
+        /// The publisher should start flushing when the batch size has been exceeded
+        #[test]
+        fn batching_batches() {
+            let topic = "test_topic";
+            let batch_size = 3;
+            let publisher = MockPublisher::new();
+            let sink = publisher_sink(publisher, batch_size);
+            pin_mut!(sink);
+
+            // sanity check
+            assert!(sink.publisher.messages().is_empty());
```

Comment:
Why not use the `is_empty` method on the publisher itself?

Reply:
no reason, I can switch this to the inherent `is_empty`